src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs

```diff
@@ -0,0 +1,1036 @@
+                else
+                {
+                    Task writePacketTask = WritePacket(TdsEnums.HARDFLUSH);
+                    if (writePacketTask == null)
+                    {
+                        HasPendingData = true;
+                        _messageStatus = 0;
+                        return null;
+                    }
+                    else
+                    {
+                        return AsyncHelper.CreateContinuationTaskWithState(
+                            task: writePacketTask,
+                            state: this,
+                            onSuccess: static (object state) =>
+                            {
+                                TdsParserStateObject stateObject = (TdsParserStateObject)state;
+                                stateObject.HasPendingData = true;
+                                stateObject._messageStatus = 0;
+                            }
+                        );
+                    }
+                }
+            }
+        }
+
+        // Processes the tds header that is present in the buffer
+        internal bool TryProcessHeader()
+        {
+            Debug.Assert(_inBytesPacket == 0, "there should not be any bytes left in packet when ReadHeader is called");
+
+            // if the header splits buffer reads - special case!
+            if ((_partialHeaderBytesRead > 0) || (_inBytesUsed + _inputHeaderLen > _inBytesRead))
+            {
+                // VSTS 219884: when some kind of MITM (man-in-the-middle) tool splits the network packets, the message header can be split over
+                // several network packets.
+                // Note: cannot use ReadByteArray here since it uses _inBytesPacket which is not set yet.
+                do
+                {
+                    int copy = Math.Min(_inBytesRead - _inBytesUsed, _inputHeaderLen - _partialHeaderBytesRead);
+                    Debug.Assert(copy > 0, "ReadNetworkPacket read empty buffer");
+
+                    Buffer.BlockCopy(_inBuff, _inBytesUsed, _partialHeaderBuffer, _partialHeaderBytesRead, copy);
+                    _partialHeaderBytesRead += copy;
+                    _inBytesUsed += copy;
+
+                    Debug.Assert(_partialHeaderBytesRead <= _inputHeaderLen, "Read more bytes for header than required");
+                    if (_partialHeaderBytesRead == _inputHeaderLen)
+                    {
+                        // All read
+                        _partialHeaderBytesRead = 0;
+                        _inBytesPacket = ((int)_partialHeaderBuffer[TdsEnums.HEADER_LEN_FIELD_OFFSET] << 8 |
+                                  (int)_partialHeaderBuffer[TdsEnums.HEADER_LEN_FIELD_OFFSET + 1]) - _inputHeaderLen;
+
+                        _messageStatus = _partialHeaderBuffer[1];
+                        _spid = _partialHeaderBuffer[TdsEnums.SPID_OFFSET] << 8 |
+                                  _partialHeaderBuffer[TdsEnums.SPID_OFFSET + 1];
+#if !NETFRAMEWORK
```

Comment:
This is just event logging and i think it should be safe to unconditionally include this in all builds.